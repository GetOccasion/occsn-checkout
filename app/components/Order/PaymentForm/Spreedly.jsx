import React, { PureComponent } from 'react';

import _ from 'underscore';

import occsn from '../../../libs/Occasion';

import SpreedlyAPI from 'spreedly';

import { Col, FormGroup, Input, Label, Row } from 'reactstrap';

import PaymentServiceProvider from './PaymentServiceProvider';

export default class Spreedly extends PaymentServiceProvider {
  constructor() {
    super();

    _.bindAll(this,
      'handleChange'
    );

    this.state = {
      month: null,
      full_name: null,
      year: null
    };
  }

  // Initializes the iFrame using the global SpreedlyAPI object imported as a separate script
  // @note Called on componentDidMount
  // @see PaymentServiceProvider#componentDidMount
  initializeForm() {
    SpreedlyAPI.init(global.OCCSN.spreedly_key, {
      "numberEl": "spreedly-number",
      "cvvEl": "spreedly-cvv"
    });

    var defaultInputStyle = 'display: block;' +
      '  width: 80%;' +
      '  padding: 0.375rem 0.75rem;' +
      '  font-size: 1rem;' +
      '  line-height: 1.5;' +
      '  color: #495057;' +
      '  background-color: #fff;' +
      '  background-clip: padding-box;' +
      '  border: 1px solid #ced4da;' +
      '  border-radius: 0.25rem;' +
      '  transition: border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out';

    var focusInputStyle = 'color: #495057;' +
      '  background-color: #fff;' +
      '  border-color: #80bdff;' +
      '  outline: 0;' +
      '  box-shadow: 0 0 0 0.2rem rgba(0, 123, 255, 0.25)';

    SpreedlyAPI.on("ready", function () {
      SpreedlyAPI.setFieldType("number", "text");
      SpreedlyAPI.setNumberFormat("prettyFormat");
      SpreedlyAPI.setPlaceholder("number", "•••• •••• •••• ••••");
      SpreedlyAPI.setPlaceholder("cvv", "•••");
      SpreedlyAPI.setStyle("number", defaultInputStyle);
      SpreedlyAPI.setStyle("cvv", defaultInputStyle);
    });

    SpreedlyAPI.on('fieldEvent', function(name, type, activeEl, inputProperties) {
      if(type == 'focus'){
        SpreedlyAPI.setStyle(name, focusInputStyle);
      }

      if(type == 'blur'){
        SpreedlyAPI.setStyle(name, defaultInputStyle);
      }
    });

    SpreedlyAPI.on('errors', function(errors) {
    });

    SpreedlyAPI.on('paymentMethod', (token) => {
      this.paymentMethodDeferred.resolve(occsn.CreditCard.build({ id: token }));
    });
  }

  // Triggers paymentMethod event
  tokenizePaymentMethodData() {
    SpreedlyAPI.tokenizeCreditCard(this.state);
  }

  handleChange(name, e) {
    this.setState({
      [name]: e.target.value
    })
  }

  render() {
    return <section>
      <FormGroup>
        <label>Name On Card</label>
        <Input type="text" id="full_name" name="full_name" placeholder="Name On Card" onChange={ (e) => this.handleChange('full_name', e) } />
      </FormGroup>

      <FormGroup>
        <Label>Credit Card Number</Label>
        <div id="spreedly-number" style={{ height: '48px' }}></div>
      </FormGroup>

      <FormGroup>
        <Row>
          <Col xs="6">
            <Label>Expiration Date</Label>
            <Row>
              <Col xs="6">
                <Input type="text" id="month" name="month" maxlength="2" placeholder="MM" onChange={ (e) => this.handleChange('month', e) } />
              </Col>
              <Col xs="6">
                <Input type="text" id="year" name="year" maxlength="4" placeholder="YYYY" onChange={ (e) => this.handleChange('year', e) } />
              </Col>
            </Row>
          </Col>
          <Col xs="3">
            <Label>CVV</Label>
            <div id="spreedly-cvv" style={{ height: '48px' }}></div>
          </Col>
        </Row>
      </FormGroup>
    </section>;
  }
}
