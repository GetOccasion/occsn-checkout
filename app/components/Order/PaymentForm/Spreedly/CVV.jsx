import React from 'react';

import { Label } from 'reactstrap';

export default class CardNumber extends React.Component {
  shouldComponentUpdate() {
    return false;
  }

  render() {
    return <div>
      <Label>CVV</Label>
      <div id="spreedly-cvv" style={{ height: '52px' }}></div>
    </div>;
  }
}
