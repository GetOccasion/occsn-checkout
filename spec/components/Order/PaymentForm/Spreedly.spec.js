import React from 'react';

import Spreedly from 'app/components/Order/PaymentForm/Spreedly';

import SpreedlyAPI from 'spreedly';

import axios from 'axios';
import occsn from 'app/libs/Occasion';

describe('Order', () => {
  describe('PaymentForm', () => {
    describe('Spreedly', () => {
      let wrapper;

      beforeEach(() => {
        wrapper = mount(<Spreedly></Spreedly>);
      });

      afterEach(async () => {
        axios.reset();
      });

      it('renders fields', () => {
        expect(wrapper).toMatchSnapshot();
      });

      context('when nameOnCard input changed', () => {
        beforeEach(() => {
          wrapper.find('input#nameOnCard').simulate('change', { target: { value: 'J Smith' } });
        });

        it('changes nameOnCard in state', () => {
          expect(wrapper).toHaveState('nameOnCard', 'J Smith');
        });
      });

      context('when month input changed', () => {
        beforeEach(() => {
          wrapper.find('input#month').simulate('change', { target: { value: '11' } });
        });

        it('changes month in state', () => {
          expect(wrapper).toHaveState('month', '11');
        });
      });

      context('when year input changed', () => {
        beforeEach(() => {
          wrapper.find('input#year').simulate('change', { target: { value: '2019' } });
        });

        it('changes year in state', () => {
          expect(wrapper).toHaveState('year', '2019');
        });
      });

      describe('tokenizePaymentMethodData', () => {
        beforeEach(() => {
          wrapper.setState({
            month: '11',
            year: '2019',
            nameOnCard: 'J Smith'
          });

          wrapper.instance().tokenizePaymentMethodData();
        });

        it('calls Spreedly.tokenizeCreditCardData with month and year', () => {
          expect(SpreedlyAPI.tokenizeCreditCard.mock.calls[0]).toEqual([{
            month: '11',
            year: '2019',
            nameOnCard: 'J Smith'
          }])
        })
      });
    });
  });
});
