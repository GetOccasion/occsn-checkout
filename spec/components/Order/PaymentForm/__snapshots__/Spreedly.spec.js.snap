// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order PaymentForm Spreedly renders fields 1`] = `
<Spreedly>
  <section>
    <FormGroup
      tag="div"
    >
      <div
        className="form-group"
      >
        <label>
          Name On Card
        </label>
        <Input
          id="full_name"
          name="full_name"
          onChange={[Function]}
          placeholder="Name On Card"
          type="text"
        >
          <input
            className="form-control"
            id="full_name"
            name="full_name"
            onChange={[Function]}
            placeholder="Name On Card"
            type="text"
          />
        </Input>
      </div>
    </FormGroup>
    <FormGroup
      tag="div"
    >
      <div
        className="form-group"
      >
        <Label
          tag="label"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          <label
            className=""
          >
            Credit Card Number
          </label>
        </Label>
        <div
          id="spreedly-number"
          style={
            Object {
              "height": "48px",
            }
          }
        />
      </div>
    </FormGroup>
    <FormGroup
      tag="div"
    >
      <div
        className="form-group"
      >
        <Row
          tag="div"
        >
          <div
            className="row"
          >
            <Col
              tag="div"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
              xs="6"
            >
              <div
                className="col-6"
              >
                <Label
                  tag="label"
                  widths={
                    Array [
                      "xs",
                      "sm",
                      "md",
                      "lg",
                      "xl",
                    ]
                  }
                >
                  <label
                    className=""
                  >
                    Expiration Date
                  </label>
                </Label>
                <Row
                  tag="div"
                >
                  <div
                    className="row"
                  >
                    <Col
                      tag="div"
                      widths={
                        Array [
                          "xs",
                          "sm",
                          "md",
                          "lg",
                          "xl",
                        ]
                      }
                      xs="6"
                    >
                      <div
                        className="col-6"
                      >
                        <Input
                          id="month"
                          maxlength="2"
                          name="month"
                          onChange={[Function]}
                          placeholder="MM"
                          type="text"
                        >
                          <input
                            className="form-control"
                            id="month"
                            maxlength="2"
                            name="month"
                            onChange={[Function]}
                            placeholder="MM"
                            type="text"
                          />
                        </Input>
                      </div>
                    </Col>
                    <Col
                      tag="div"
                      widths={
                        Array [
                          "xs",
                          "sm",
                          "md",
                          "lg",
                          "xl",
                        ]
                      }
                      xs="6"
                    >
                      <div
                        className="col-6"
                      >
                        <Input
                          id="year"
                          maxlength="4"
                          name="year"
                          onChange={[Function]}
                          placeholder="YYYY"
                          type="text"
                        >
                          <input
                            className="form-control"
                            id="year"
                            maxlength="4"
                            name="year"
                            onChange={[Function]}
                            placeholder="YYYY"
                            type="text"
                          />
                        </Input>
                      </div>
                    </Col>
                  </div>
                </Row>
              </div>
            </Col>
            <Col
              tag="div"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
              xs="3"
            >
              <div
                className="col-3"
              >
                <Label
                  tag="label"
                  widths={
                    Array [
                      "xs",
                      "sm",
                      "md",
                      "lg",
                      "xl",
                    ]
                  }
                >
                  <label
                    className=""
                  >
                    CVV
                  </label>
                </Label>
                <div
                  id="spreedly-cvv"
                  style={
                    Object {
                      "height": "48px",
                    }
                  }
                />
              </div>
            </Col>
          </div>
        </Row>
      </div>
    </FormGroup>
  </section>
</Spreedly>
`;
